### DISCUSSION TITLE
Use my own SSL-Certificates and not Let's Encrypt?

### DISCUSSION THREAD

**Discussion by hspies91** [requarks/wiki#5179] [Help / Questions]
Hey Guys, 

i installed Wiki JS following this Article: 

https://docs.requarks.io/install/ubuntu and started to use the wiki for my own internal stuff. 
Now, i wanted to use https and publish the Server BUT i can not use LE for several reasons (no forwarding of Port 80 etc.). 
How can i now use HTTPS and pass my Cert & Key File in this Line? 

docker create --name=wiki -e "HERE NOT LE, I WANT TO PASS MY OWN CERTIFICATES" -e SSL_ACTIVE=1 -e DB_TYPE=postgres -e DB_HOST=db -e DB_PORT=5432 -e DB_PASS_FILE=/etc/wiki/.db-secret -v /etc/wiki/.db-secret:/etc/wiki/.db-secret:ro -e DB_USER=wiki -e DB_NAME=wiki -e UPGRADE_COMPANION=1 --restart=unless-stopped -h wiki --network=wikinet -p 80:3000 -p 443:3443 ghcr.io/requarks/wiki:2

Is this possible?
Best Regards, :)

**Comment by eglyn** [comment]
I up this thread, as I have the same question :)
I find some variables in the request section: SSL_PEM_CERT_FILEPATH and SSL_PEM_KEY_FILEPATH, but it does not work. It seems wiki looks always for a letsencrypt email, thx !

`docker create --name=wiki -e SSL_PROVIDER=CUSTOM -e SSL_PEM_CERT_FILEPATH=/opt/ssl/domain.cert.pem -e SSL_PEM_KEY_FILEPATH=/opt/ssl/domain.key.pem -e SSL_ACTIVE=1 -e DB_TYPE=postgres -e DB_HOST=db -e DB_PORT=5432 -e DB_PASS_FILE=/etc/wiki/.db-secret -v /etc/wiki/.db-secret:/etc/wiki/.db-secret:ro -e DB_USER=wiki -e DB_NAME=wiki -e UPGRADE_COMPANION=1 --restart=unless-stopped -h wiki --network=wikinet -p 80:3000 -p 443:3443 ghcr.io/requarks/wiki:2`

![image](https://user-images.githubusercontent.com/24936778/175250719-9023fdaf-79ed-43ea-8db8-61d3656d5633.png)

**Comment by eglyn** [comment]
Ok, i answer on my own question :D I use a apache proxy, and it works great, my conf: 

- wiki only on port 3000 http
`docker create --name=wiki -e DB_TYPE=postgres -e DB_HOST=db -e DB_PORT=5432 -e DB_PASS_FILE=/etc/wiki/.db-secret -v /etc/wiki/.db-secret:/etc/wiki/.db-secret:ro -e DB_USER=wiki -e DB_NAME=wiki -e UPGRADE_COMPANION=1 --restart=unless-stopped -h wiki --network=wikinet -p 3000:3000 ghcr.io/requarks/wiki:2`

- install apache
`apt install apache2`
`a2enmod ssl rewrite headers proxy proxy_http`

- config apache: 
```
<VirtualHost *:80>
        ServerName myserver.mydomain.com
        RewriteEngine On
        RewriteRule ^(.*)$ https://%{HTTP_HOST}$1 [redirect=301]


</VirtualHost>

<IfModule mod_ssl.c>
        <VirtualHost *:443>
                ServerName myserver.mydomain.com
                # Use HTTP Strict Transport Security to force client to use secure connections only.
                Header always set Strict-Transport-Security "max-age=31536000; includeSubDomains; preload"
                # Referer logging is used to allow websites and web servers to identify where people are visiting them from, for promotional or statistical purposes.
                Header always set Referrer-Policy "no-referrer"

                SSLCertificateFile /opt/ssl/myserver.mydomain.com.cer
                SSLCertificateKeyFile /opt/ssl/myserver.mydomain.com.fr.key

                SSLEngine on
                SSLProtocol all -TLSv1 -TLSv1.1 -SSLv2 -SSLv3
                SSLHonorCipherOrder on
                SSLCompression off
                SSLOptions +StrictRequire
                SSLCipherSuite ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA>

                ProxyPreserveHost On
                ProxyPass / http://localhost:3000/
                ProxyPassReverse / http://localhost:3000/
        </VirtualHost>
</IfModule>
```

**Reply by Balkaran-Singh-Gabby** [reply]
Thanks, I workes great! Apache is the best...
I was wasting lot of time following official documentation...

**Comment by riversdev0** [comment]
For anyone else who ends up here (like I did), there is a mention under the HTTPS header on the [installation page](https://docs.requarks.io/install/docker#https) for software version 2 that says this:
> If you want to provide your own SSL certificate configuration, you must instead mount a config file [as explained below](https://docs.requarks.io/install/docker#alternative-mount-the-config-file).

There is a sample config.yml available [on GitHub](https://github.com/Requarks/wiki/blob/master/config.sample.yml).

So, I removed all my environment variables from my Docker command, created a config.yml on my Docker host instead, and then mapped my config.yml into the container in the `/wiki` folder, by adding this to my Docker command:
```
-v /home/user/wiki.js/config.yml:/etc/config.yml
```
And now it works!

**Comment by harvthe2** [comment]
On our RHEL8 install I placed the certificate and key in the wiki certificates folder (/var/wikijs/certificates/ on my particular instance)
And Administration -> SSL -> Redirect HTTP to HTTPS and click TURN ON 
Running wikijs 2.5.307

**Comment by bkelly1984** [comment]
Hey fellow Wiki.js users,

I ran into this issue as well, specifically with DigitalOcean’s Wiki.js droplet application, where the above solutions didn’t quite apply. That said, posts by @eglyn, @riversdev0, and @harvthe2 pointed me in the right direction and helped me discover that Let’s Encrypt is hard-coded as the default in this setup.

From _/wiki/config.yml_ inside _ghcr.io/requarks/wiki:2_:
```
ssl:
  enabled: $(SSL_ACTIVE)
  port: 3443
  provider: letsencrypt
  domain: $(LETSENCRYPT_DOMAIN)
  subscriberEmail: $(LETSENCRYPT_EMAIL)

```
To use a custom certificate, I replaced this with my own _config.yml_. Below is the process I followed, using _myserver_ as a stand-in name.

### 1) Copy your custom certificate files
```
cp myserver.key /etc/wiki/
cp myserver.crt /etc/wiki/
cp myserver.ca-bundle /etc/wiki/
chmod 444 /etc/wiki/myserver.*
```
### 2) Create a custom _config.yml_

This overrides the Let’s Encrypt settings and uses a custom PEM certificate.
```
cat > /etc/wiki/config.yml << EOF
port: 3000
bindIP: 0.0.0.0

db:
  type: \$(DB_TYPE)
  host: '\$(DB_HOST)'
  port: \$(DB_PORT)
  user: '\$(DB_USER)'
  pass: '\$(DB_PASS)'
  db: \$(DB_NAME)
  storage: \$(DB_FILEPATH)
  ssl: \$(DB_SSL)

ssl:
  enabled: \$(SSL_ACTIVE)
  port: 3443
  provider: custom
  format: pem
  key: /etc/wiki/myserver.key
  cert: /etc/wiki/myserver.crt
  passphrase: null
  dhparam: null

logLevel: \$(LOG_LEVEL:info)
logFormat: \$(LOG_FORMAT:default)
ha: \$(HA_ACTIVE)
EOF
```
### 3) Stop and remove the existing Wiki.js container
```
docker stop wiki
docker rm wiki
```
### 4) Required environment variables
Add new environment variables. NODE_EXTRA_CA_CERTS is required so Node trusts your custom CA bundle:
```
-e SSL_ACTIVE=1
-e NODE_EXTRA_CA_CERTS=/etc/wiki/myserver.ca-bundle
```
### 5) Volume mappings

Add these volume mounts so the container can access your files:

```
-v /etc/wiki/config.yml:/wiki/config.yml:ro
-v /etc/wiki/myserver.key:/etc/wiki/myserver.key:ro
-v /etc/wiki/myserver.crt:/etc/wiki/myserver.crt:ro
-v /etc/wiki/myserver.ca-bundle:/etc/wiki/myserver.ca-bundle:ro
```
### 6) Create the container

Here’s my full docker create command (adjust as needed):

```
docker create \
  --name=wiki \
  -e SSL_ACTIVE=1 \
  -e NODE_EXTRA_CA_CERTS=/etc/wiki/myserver.ca-bundle \
  -e DB_TYPE=postgres \
  -e DB_HOST=db \
  -e DB_PORT=5432 \
  -e DB_USER=wiki \
  -e DB_NAME=wiki \
  -e DB_PASS_FILE=/etc/wiki/.db-secret \
  -e UPGRADE_COMPANION=1 \
  -v /etc/wiki/.db-secret:/etc/wiki/.db-secret:ro \
  -v /etc/wiki/config.yml:/wiki/config.yml:ro \
  -v /etc/wiki/myserver.key:/etc/wiki/myserver.key:ro \
  -v /etc/wiki/myserver.crt:/etc/wiki/myserver.crt:ro \
  -v /etc/wiki/myserver.ca-bundle:/etc/wiki/myserver.ca-bundle:ro \
  --restart=unless-stopped \
  -h wiki \
  --network=wikinet \
  -p 80:3000 \
  -p 443:3443 \
  ghcr.io/requarks/wiki:2

```

Hope this saves someone else a few hours of digging.
